tardis/opacities/macro_atom/base.py

```diff
@@ -9,7 +9,231 @@
+    beta_sobolev,
+    j_blues,
+    stimulated_emission_factor,
+    transition_probability_coef,
+    block_references,
+    normalize,
+):
+    """Calculate the non-markov transition probabilities using optimized functions
+    Parameters
+    ----------
+    macro_atom_data : pd.DataFrame
+        Macro Atom Data
+    beta_sobolev : pd.DataFrame
+        Beta Sobolevs
+    j_blues : pd.DataFrame
+        mean intensity
+    stimulated_emission_factor : np.ndarray
+        Stimulated Emission Factors
+    transition_probability_coef : np.ndarray
+        Reshaped macro atom transition probabilities
+    block_references : np.ndarray
+        macro atom block references
+    normalize : bool
+        Whether or not to normalize the transition probabilities to unity
+
+    Returns
+    -------
+    np.ndarray
+        transition probabilities
+    """
+    transition_probabilities = np.empty(
+        (transition_probability_coef.shape[0], beta_sobolev.shape[1])
+    )
+    # trans_old = self.calculate_transition_probabilities(macro_atom_data, beta_sobolev, j_blues, stimulated_emission_factor)
+    transition_type = macro_atom_data.transition_type.values
+    lines_idx = macro_atom_data.lines_idx.values
+    tpos = macro_atom_data.transition_probability.values
+    util.fast_calculate_transition_probabilities(
+        tpos,
+        beta_sobolev.values,
+        j_blues.values,
+        stimulated_emission_factor,
+        transition_type,
+        lines_idx,
+        block_references,
+        transition_probabilities,
+        normalize,
+    )
+    return transition_probabilities
+
+
+def initialize_macro_atom_transition_type_filters(atomic_data, macro_atom_data):
+    """Get the filters and block references from the macro atom
+
+    Parameters
+    ----------
+    atomic_data : tardis.io.atom_data.AtomData
+            Atomic Data
+    macro_atom_data : pd.DataFrame
+        Macro Atom Data
+
+    Returns
+    -------
+    np.ndarray
+        Mask where the transition type is 1
+    np.ndarray
+        index of lines at these locations
+    pd.ndarray
+        macro atom block references
+    """
+    transition_up_filter = macro_atom_data.transition_type.values == 1
+    transition_up_line_filter = macro_atom_data.lines_idx.values[
+        transition_up_filter
+    ]
+    block_references = np.hstack(
+        (
+            atomic_data.macro_atom_references.block_references,
+            len(macro_atom_data),
+        )
+    )
+
+    return transition_up_filter, transition_up_line_filter, block_references
+
+
+def get_transition_probability_coefs(macro_atom_data):
+    """Coefficients of the transition probabilities
+
+    Parameters
+    ----------
+    macro_atom_data : pd.DataFrame
+        Macro Atom Data
+
+    Returns
+    -------
+    np.ndarray
+        Reshaped macro atom transition probabilities
+    """
+    return macro_atom_data.transition_probability.values[np.newaxis].T
+
+
+def get_macro_atom_data(atomic_data):
```

Comment:
I think this shows us that we don't need *all* the atomic data to be passed around all the time.

Reply:
Definitely not, some kind of MacroAtomData structure could be useful